pkg/volume/cinder/attacher.go

```diff
@@ -17,191 +17,86 @@ limitations under the License.
 package cinder
 
 import (
-	"fmt"
 	"os"
-	"path"
-	"strings"
 	"time"
-
-	"github.com/golang/glog"
-	"k8s.io/apimachinery/pkg/types"
+	
+  "k8s.io/apimachinery/pkg/types"
 	"k8s.io/kubernetes/pkg/util/exec"
 	"k8s.io/kubernetes/pkg/util/mount"
 	"k8s.io/kubernetes/pkg/volume"
-	volumeutil "k8s.io/kubernetes/pkg/volume/util"
 )
 
 type cinderDiskAttacher struct {
-	host           volume.VolumeHost
-	cinderProvider CinderProvider
-}
-
-var _ volume.Attacher = &cinderDiskAttacher{}
-
-var _ volume.AttachableVolumePlugin = &cinderPlugin{}
-
-const (
-	checkSleepDuration = time.Second
-)
-
-func (plugin *cinderPlugin) NewAttacher() (volume.Attacher, error) {
-	cinder, err := getCloudProvider(plugin.host.GetCloudProvider())
-	if err != nil {
-		return nil, err
-	}
-	return &cinderDiskAttacher{
-		host:           plugin.host,
-		cinderProvider: cinder,
-	}, nil
-}
-
-func (plugin *cinderPlugin) GetDeviceMountRefs(deviceMountPath string) ([]string, error) {
-	mounter := plugin.host.GetMounter()
-	return mount.GetMountRefs(mounter, deviceMountPath)
+	*cinderVolume
 }
 
 func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
-	volumeSource, _, err := getVolumeSource(spec)
-	if err != nil {
-		return "", err
-	}
-
-	volumeID := volumeSource.VolumeID
-
-	instances, res := attacher.cinderProvider.Instances()
-	if !res {
-		return "", fmt.Errorf("failed to list openstack instances")
-	}
-	instanceid, err := instances.InstanceID(nodeName)
+	manager, err := getManager(attacher.plugin, spec, nil)
```

Comment:
The Deleter and the Provisioner caches the `manager` inside `cinderVolume` as their constructors have either `volume.Spec` or `volume.VolumeOptions`, letting us determine which manager should be used right away (their instances are volume specific). Unfortunately, the Attacher and Detacher only get `volume.Spec` during the actual method calls and we therefore need to determine the manager at this time (their instance are not constructed as volume specific - even though it might be the case in practice).